src/Homeworks/Homework_6/AVL_Tree.py

```diff
@@ -0,0 +1,235 @@
+            current_root = current_root.right
+        else:
+            current_root = current_root.left
+    return None
+
+
+def get(map: Tree, key: Any) -> Any:
+    node = get_node(map, key)
+    if node is not None:
+        return node.value
+    raise ValueError("BST hasn't Node with this key")
+
+
+def has_key(map: Tree, key: Any) -> bool:
+    return get_node(map, key) is not None
+
+
+def get_maximum(map: Tree) -> Any:
+    current_root = map.root
+    maximum_key = current_root.key
+    while current_root is not None:
+        maximum_key = current_root.key
+        current_root = current_root.right
+    return maximum_key
+
+
+def get_minimum(map: Tree) -> Any:
+    current_root = map.root
+    minimum_key = current_root.key
+    while current_root is not None:
+        minimum_key = current_root.key
+        current_root = current_root.left
+    return minimum_key
+
+
+def get_upper_bound(map: Tree, key: Any) -> Any:
+    if get_maximum(map) < key:
+        raise ValueError("In the map there not key bigger than given")
+    current_root = map.root
+    upper_bound = current_root.key
+    while current_root is not None:
+        upper_bound = current_root.key if current_root.key > key else upper_bound
+        if current_root.key > key:
+            current_root = current_root.left
+        else:
+            current_root = current_root.right
+    return upper_bound
+
+
+def get_lower_bound(map: Tree, key: Any) -> Any:
+    if has_key(map, key):
+        return key
+    return get_upper_bound(map, key)
+
+
+def find_min_in_right_subtree(root: TreeNode) -> TreeNode:
+    current_root = root.right
+    if current_root is None:
+        return root
+    while current_root.left is not None:
+        current_root = current_root.left
+    return current_root
+
+
+def remove(map: Tree, key: Any) -> Any:
+    if not has_key(map, key):
+        raise ValueError("BST hasn't Node with this key")
+    deleting_value = get(map, key)
+
+    def _remove(current_root: TreeNode, key: Any):
+        if current_root.key > key:
+            current_root.left = _remove(current_root.left, key)
+        elif current_root.key < key:
+            current_root.right = _remove(current_root.right, key)
+        else:
+            if current_root.left is None:
+                return current_root.right
+            elif current_root.right is None:
+                return current_root.left
+
+            new_node = find_min_in_right_subtree(current_root)
+            current_root.key, current_root.value = new_node.key, new_node.value
+            current_root.right = _remove(current_root.right, current_root.key)
+        update_height(current_root)
+        rotate_root = balancing(current_root, key)
+        if rotate_root is not None:
+            return rotate_root
+        return current_root
+
+    map.root = _remove(map.root, key)
+    return deleting_value
+
+
+def pre_order_traverse(root: TreeNode, nodes):
+    if not (root is None):
```

Comment:
```suggestion
    if root is not None:
```
и в остальных функциях ниже также